File: Javascript/DFS/M-841. Keys and Rooms.ts
function canVisitAllRooms(rooms: number[][]): boolean {
  const keys = new Set<number>([0]);

  const dfs = (room: number) => {
    for (const key of rooms[room]) {
      if (keys.has(key)) continue;
      keys.add(key);
      dfs(key);
    }
  };

  dfs(0);
  return keys.size === rooms.length;
}

const testData = [
  {
    rooms: [[1], [2], [3], []],
    ans: true,
  },
  {
    rooms: [[1, 3], [3, 0, 1], [2], [0]],
    ans: false,
  },
];

for (const data of testData) {
  console.log(data['ans'] === canVisitAllRooms(data['rooms']));
}

File: DFS/M-841. Keys and Rooms/M-841. Keys and Rooms.py
class Solution:
    @classmethod
    def canVisitAllRooms(self, rooms: list[list[int]]) -> bool:
        keys = set([0])

        def dfs(room: int) -> None:
            for key in rooms[room]:
                if key in keys:
                    continue
                keys.add(key)
                dfs(key)

        dfs(0)
        return len(keys) == len(rooms)


testData = [
    {
        "rooms": [[1], [2], [3], []],
        "ans": True,
    },
    {
        "rooms": [[1, 3], [3, 0, 1], [2], [0]],
        "ans": False,
    },
]

for data in testData:
    print(data["ans"] == Solution.canVisitAllRooms(data["rooms"]))
